Infinite steams cookbook recipe

LaurentRDC opened the pull request:
This is a cookbook recipe to serve infinite HTTP streams. There are a lot of subtleties in providing an HTTP infinite stream because data must be sent to the client on a regular basis.

LaurentRDC committed:
Infinite steams cookbook recipe

LaurentRDC committed:
Description

LaurentRDC commented:
https://github.com/kazu-yamamoto/crypton-certificate/issues/21

tchoutri commented:
@LaurentRDC The package's metadata seems to have been amended on Hackage. :)

tchoutri commented on doc/cookbook/infinite-streams/InfiniteStreams.lhs:
```diff
@@ -0,0 +1,277 @@
+
+    -- Writer thread will feed our Chan forever. This is where
+    -- your secret random number generation algorithm would go.
+    -- Unfortunately, we have some technical debt here.
```
> Unfortunately, we have some technical debt here.

What is the intended purpose of this comment?

LaurentRDC replied:
It was meant as a joke, as the "random numbers" are not random at all. I'll remove the comment

tchoutri commented on doc/cookbook/infinite-streams/InfiniteStreams.lhs:
````diff
@@ -0,0 +1,277 @@
+```
+
+We start with our scenario: we are tasked with creating an API which will serve random numbers in real-time. We are
+given a function that creates a producer of characters, and a method for us to stop the producer:
````
> a producer of characters

I thought we were producing an infinite stream of number. There does not seem to be any further mention of characters after that, have I missed something?

LaurentRDC commented:
Thanks for your review!

While addressing your comments, I have also refactored the tricky logic that interleaves integers with `KeepAlive` messages to use `withAsync`. I find this more intuitive, and easier to explain! Let me know what you think.

LaurentRDC commented:
> What is the reasoning behind the use of throwIO in favour of error? Do you get nicer callstacks or something?

There's no deep reason. I personally think of `error` as "quick-and-dirty`, and 'throwIO` as more proper.

tchoutri reviewed (approved):
Perfect, let's go!

LaurentRDC commented:
Thanks for reviewing @tchoutri !

LaurentRDC commented:
Very interesting! Thanks for sharing!
So basically, `error` *is* quick and dirty compared to `throwIO`.
